sdk/canton/community/participant/src/main/scala/com/digitalasset/canton/participant/admin/PackageUpgradeValidator.scala

```diff
@@ -8,35 +8,80 @@ import com.daml.daml_lf_dev.DamlLf.Archive
 import com.daml.error.DamlError
 import com.daml.lf.archive.Decode
 import com.daml.lf.data.Ref
-import com.daml.lf.language.Ast
+import com.daml.lf.language.Util.dependenciesInTopologicalOrder
+import com.daml.lf.language.{Ast, LanguageVersion}
 import com.daml.lf.validation.{TypecheckUpgrades, UpgradeError}
 import com.daml.logging.entries.LoggingValue.OfString
 import com.digitalasset.canton.ledger.error.PackageServiceErrors.{InternalError, Validation}
 import com.digitalasset.canton.logging.LoggingContextWithTrace.implicitExtractTraceContext
 import com.digitalasset.canton.logging.{LoggingContextWithTrace, NamedLoggerFactory, NamedLogging}
+import com.digitalasset.canton.participant.admin.PackageUpgradeValidator.PackageMap
 import com.digitalasset.canton.participant.admin.PackageUploader.ErrorValidations
 import com.digitalasset.canton.tracing.TraceContext
-import scalaz.std.either.*
-import scalaz.std.option.*
+import com.digitalasset.canton.util.EitherTUtil
+import scalaz.std.either._
+import scalaz.std.option._
 import scalaz.std.scalaFuture.futureInstance
-import scalaz.syntax.traverse.*
+import scalaz.syntax.traverse._
 
 import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits.infixOrderingOps
+
+object PackageUpgradeValidator {
+  type PackageMap = Map[Ref.PackageId, (Ref.PackageName, Ref.PackageVersion)]
+}
 
 class PackageUpgradeValidator(
-    getPackageMap: TraceContext => Map[Ref.PackageId, (Ref.PackageName, Ref.PackageVersion)],
+    getPackageMap: TraceContext => PackageMap,
     getLfArchive: TraceContext => Ref.PackageId => Future[Option[Archive]],
     val loggerFactory: NamedLoggerFactory,
 )(implicit executionContext: ExecutionContext)
     extends NamedLogging {
 
   def validateUpgrade(
+      upgradingPackages: List[(Ref.PackageId, Ast.Package)]
+  )(implicit
+      loggingContext: LoggingContextWithTrace
+  ): EitherT[Future, DamlError, Unit] = {
+    val upgradingPackagesMap = upgradingPackages.toMap
+    val deps = dependenciesInTopologicalOrder(upgradingPackages.map(_._1), upgradingPackagesMap)
+    val packageMap = getPackageMap(loggingContext.traceContext)
+
+    def go(
+        packageMap: PackageMap,
+        deps: List[Ref.PackageId],
+    ): EitherT[Future, DamlError, PackageMap] = deps match {
+      case Nil => EitherT.pure[Future, DamlError](packageMap)
+      case pkgId :: rest =>
+        val pkg = upgradingPackagesMap(pkgId)
+        val supportsUpgrades = pkg.languageVersion >= LanguageVersion.Features.packageUpgrades
+        // TODO: If pkgA-V1 (1.16) and pkgA-V2 (1.16) depend on pkgB (1.15)
+        // the pkgB in V1 and V2 must have the same package id (as defined by its package name + version)
+        pkg.metadata match {
+          case Some(pkgMetadata) =>
+            for {
+              _ <- EitherTUtil.ifThenET(supportsUpgrades)(
+                validatePackageUpgrade((pkgId, pkg), pkgMetadata, packageMap)
```

Comment:
As discussed offline: this is pretty unlikely and there is no timeout so worst case it makes startup slow but doesn't prevent a participant from starting. I've added a comment explaining that this could happen and why we haven't added a cache.